### PR TITLE
Added backend socket

### PR DESCRIPTION
**Pull request by Hwid2121**
the commit of all the modifies are in the branch back_end_socekts

This branch created for merge all backend modifies inside of develop

@micheledallerive

### CODE REVIEW

File: serverSocket.js [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    joinChat(userChat, socket);
+    socket.emit('chats:read', userChat);
+
+    socket.on('disconnect', () => {
+      console.log('⛔User disconnected with id ' + socket.id);
+    });
+  });
+}
+
+async function getChats(userId) {
```
Comment (JekxDevil) [inline serverSocket.js:36]: declare functions only used in init e.g. `getChats` inside init.

File: routes/chat.js [hunk 2]
```diff
@@ -133,6 +140,9 @@ router.post('/chat/:chatId/messages', async function (req, res) {
     delivered_at: new Date(),
   });
 
+  // socket io emit to all users in the chat room that a new message has been created and send the new message
+  io.to(req.params.chatId).emit('messages:create', newMessage);
```
Comment (JekxDevil) [inline routes/chat.js:144]: create a function with such functionality in `serverSocket.js` and export it.
Trying to keep the socket functionality in `serverSocket.js` and use it elsewhere with encapsulating functions

File: serverSocket.js [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    .populate('chat')
+    .exec();
+  if (error) {
+    console.log(error);
+    return [];
+  }
+  return userChats.map((userChat) => userChat.chat);
+}
+
+function joinChat(userChats, socket) {
```
Comment (JekxDevil) [inline serverSocket.js:47]: parametrize `joinChat` to accept a userId and join the socket of the user (so the room corresponding to the Id) to a chatId.
Then export the function to use it elsewhere, e.g. `chat.js` 87-89 .

File: routes/chat.js [hunk 4]
```diff
@@ -84,6 +84,13 @@ router.post('/chat', async function (req, res) {
       chat: chatId,
     });
 
+    const userSocket = [io.sockets.adapter.rooms.get(req.userId)];
```
Comment (JekxDevil) [inline routes/chat.js:87]: see `joinChat` function comment on `serverSocket.js`

File: routes/chat.js [hunk 5]
```diff
@@ -84,6 +84,13 @@ router.post('/chat', async function (req, res) {
...
+    const otherSocket = [io.sockets.adapter.rooms.get(otherId)];
+    userSocket.join(chatId.toString());
+    otherSocket.join(chatId.toString());
+
+    io.to(req.userId).to(otherId).emit('chats:created', chat);
```
Comment (JekxDevil) [inline routes/chat.js:92]: create a function with such functionality in serverSocket.js and export it.
Trying to keep the socket functionality in serverSocket.js and use it elsewhere with encapsulating functions

File: serverSocket.js (hunk 1 above)
Comment (Hwid2121) [reply serverSocket.js]: 6be6741f522fd9de2f4990c78af52f272332778c

File: routes/chat.js (hunk 5 above)
Comment (Hwid2121) [reply routes/chat.js:94]: cc5e43e

File: serverSocket.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    joinChat(userChat, socket);
+    socket.emit('chats:read', userChat);
+
+    socket.on('disconnect', () => {
+      console.log('⛔User disconnected with id ' + socket.id);
+    });
+  });
+}
+
+function joinChat(userChats, socket) {
```
Comment (micheledallerive) [inline serverSocket.js:47]:
```suggestion
function joinRooms(rooms, socket) {
```

File: serverSocket.js [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  if (!Array.isArray(userChats)) {
+    socket.join(userChats);
+  } else {
+    userChats.forEach((chat) => {
+      socket.join(chat._id.toString());
+    });
+  }
```
Comment (micheledallerive) [inline serverSocket.js:54]:
```suggestion
    rooms.forEach((room) => {
      socket.join(room);
    });
  }
```

File: serverSocket.js [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      console.log(e);
+      next(new Error('Authentication error'));
+    }
+  });
+
+  io.on('connection', (socket) => {
+    console.log('✅User connected with id ' + socket.id);
+    console.log('Chat list is coming...');
+    socket.join(socket.userID);
+    const userChat = getChats(socket.userID);
```
Comment (micheledallerive) [inline serverSocket.js:37]:
```suggestion
    const userChats = getChats(socket.userID);
```

File: serverSocket.js [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    joinChat(userChat, socket);
```
Comment (micheledallerive) [inline serverSocket.js:38]:
```suggestion
    joinChat(userChats.map((chat)=>chat._id.toString(), socket);
```

**Review by micheledallerive — CHANGES_REQUESTED** [review]

**Review by micheledallerive — APPROVED** [review]

File: middlewares/socket.middleware.js [hunk 10]
```diff
@@ -0,0 +1,19 @@
+const { authConstants } = require('../constants/auth.constants');
+const { verifyJWT } = require('../utils/auth.utils');
+
+const authMiddleware = (socket, next) => {
```
Comment (micheledallerive) [inline middlewares/socket.middleware.js:4]: Add jsDoc

File: serverSocket.js [hunk 11]
```diff
@@ -0,0 +1,49 @@
+const io = require('socket.io')();
+const { UserChat } = require('./models/userChat');
+const { authMiddleware } = require('./middlewares/socket.middleware');
+
+function init(server) {
+  io.attach(server);
+
+  async function getChats(userId) {
```
Comment (micheledallerive) [inline serverSocket.js:8]: jsDoc

**Review by micheledallerive — CHANGES_REQUESTED** [review]

File: routes/chat.js [hunk 12]
```diff
@@ -84,6 +86,13 @@ router.post('/chat', async function (req, res) {
       chat: chatId,
     });
 
+    const userSocket = [io.sockets.adapter.rooms.get(req.userId)];
+    const otherSocket = [io.sockets.adapter.rooms.get(otherId)];
+    serverSocket.joinRooms(chatId.toString(), otherSocket);
```
Comment (DonnaDona) [inline routes/chat.js:91]: Instead of passing one user at a time, it would be more beneficial to give an array of users to be then used to create groups.

**Review by DonnaDona — CHANGES_REQUESTED** [review]

File: routes/chat.js [hunk 13]
```diff
@@ -5,6 +5,8 @@ const { ObjectId } = require('mongodb');
 const { User } = require('../models/user');
 const { Chat } = require('../models/chat');
 const { Message, MessageType } = require('../models/message'); // MessageType is used for verification
+const io = require('../serverSocket').io;
```
Comment (JekxDevil) [inline routes/chat.js:8]: really bad style to import io without incapsulating its functionality in methods.

**Review by JekxDevil — APPROVED** [review]

**Review by micheledallerive — APPROVED** [review]

File: serverSocket.js [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+function joinRooms(rooms, socket) {
+  if (!Array.isArray(rooms)) {
+    rooms = [rooms];
+  }
+  rooms.forEach((room) => {
+    socket.join(room);
+  });
+}
```
Comment (micheledallerive) [inline serverSocket.js:46]:
```suggestion
function joinRooms(rooms, sockets) {
  if (!Array.isArray(rooms)) {
    rooms = [rooms];
  }
  if (!Array.isArray(sockets)) {
  sockets = [sockets];
  }
  rooms.forEach((room) => {
    sockets.forEach((socket)=>{
      socket.join(room);
    });
  });
}
```

**Review by micheledallerive — CHANGES_REQUESTED** [review]

File: serverSocket.js [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+const { authMiddleware } = require('./middlewares/socket.middleware');
+
+// Initialize the socket.io server
+function init(server) {
+  io.attach(server);
+
+  async function getChats(userId) {
+    const [error, userChats] = await UserChat.find({ user: userId })
+      .populate('chat')
+      .exec();
```
Comment (micheledallerive) [inline serverSocket.js:12]: Populate the messages: https://mongoosejs.com/docs/populate.html#deep-populate
```suggestion
    const [error, userChats] = await UserChat.find({ user: userId })
      .populate({
        path: 'chat',
        populate: { path: 'messages' }
      })
      .exec();
```

**Review by micheledallerive — CHANGES_REQUESTED** [review]

**Review by micheledallerive — APPROVED** [review]

**Review by DonnaDona — APPROVED** [review]

**Pull request merged**